### PR TITLE
Feature/mushdoom ranged

### PR DESCRIPTION
**Pull request by JoanStinson**
What's new:
- Mushdooms have ranged behavior, they switch between if at least 3 mushdooms are alive
- A target point is placed down the player and the mushdoom shoots a ball in parabola

To test it, be sure in the scene there is a: 
- "Target Point" go from the folder "Final Mushdoom > Ball > TargetPoint prefab". Make sure it's child go is scale is (100, 1, 100) and it's emmisive is white.
- "Mushdoom Ball" go which has the script MushdoomBall, it's first go child is a sphere and it's second go child is a sphere collider. I tried to make it also a prefab but the engine kept crashing :)
It should like this: https://gyazo.com/e8d96ddff14e9be5fb28334fd4d94a7e

Expected behavior: https://gyazo.com/2b39d68a95e74c044dd168a3c0675f5f

### CODE REVIEW

File: Scripts/src/Script/EnemyController.cpp [hunk 1]
```diff
@@ -22,6 +22,7 @@
 #include "PlayerController.h"
 #include "UIManager.h"
 
+#pragma region Generic
```
Comment (JorxSS) [inline Scripts/src/Script/EnemyController.cpp:25]: why are you using this pragma?

File: Scripts/src/Script/EnemyManager.cpp [hunk 2]
```diff
@@ -92,23 +95,23 @@ void EnemyManager::SpawnEnemy(const unsigned type, const float3& spawn_position)
...
 		{
 			enemy = static_cast<Mushdoom*>(enemies[i]);
 
 			enemy->ResetEnemy();
-			
+
 			enemy->owner->transform.SetTranslation(spawn_position);
 			enemy->collider->UpdateDimensions();
-			enemy->collider->active_physics = true;
+			//enemy->collider->active_physics = true;
```
Comment (JorxSS) [inline Scripts/src/Script/EnemyManager.cpp:108]: why now the physics are not active? and if not need it you should delete this line

File: Scripts/src/Script/EnemyController.cpp (hunk 1 above)
Comment (JoanStinson) [reply Scripts/src/Script/EnemyController.cpp:25]: To make folders in the code, so I can hide groups of functions, it's getting very crowded

File: Scripts/src/Script/EnemyManager.cpp (hunk 2 above)
Comment (JoanStinson) [reply Scripts/src/Script/EnemyManager.cpp:108]: Because I already set it in the ResetEnemy method

File: Scripts/src/Script/Biter.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
+#include "Biter.h"
```
Comment (mariofv) [inline Scripts/src/Script/Biter.cpp:1]: We should avoid mixing things that are not related to the PR

File: Scripts/src/Script/EnemyController.cpp [hunk 4]
```diff
@@ -217,9 +217,53 @@ void EnemyController::BattleCircleAI()
...
+	float3 target = current_target->owner->transform.GetTranslation();
+	float3 position = owner->transform.GetTranslation();
+	float3 desired_velocity = float3::zero;
+
+	float distance = abs(position.DistanceSq(target));
+
+	// Move until enemy reaches danger distance
+	if (distance > danger_distance && !is_throwing)
+	{
+		desired_velocity = target - position;
```
Comment (mariofv) [inline Scripts/src/Script/EnemyController.cpp:235]: Is intended velocity being variable with the distance between the player and the enemy?

File: Scripts/src/Script/EnemyController.cpp [hunk 5]
```diff
@@ -240,25 +284,34 @@ void EnemyController::Seek(float3& velocity)
 		}
 		else
 		{
-			owner->transform.LookAt(position);
+			float3 look_pos = float3(current_target->owner->transform.GetTranslation().x, owner->transform.GetTranslation().y, current_target->owner->transform.GetTranslation().z);
```
Comment (mariofv) [inline Scripts/src/Script/EnemyController.cpp:287]: Why not assign here directly the vector?
`float3 look_pos = current_target->owner->transform.GetTranslation()`.
Also, we should avoid using the local translation and use the global translation

File: Scripts/src/Script/EnemyController.cpp [hunk 6]
```diff
@@ -240,25 +284,34 @@ void EnemyController::Seek(float3& velocity)
...
 	float3 perpendicular_vec = float3::unitY.Cross(target_position);
 	float3 desired_velocity = perpendicular_vec * direction;
 	Seek(desired_velocity);
 }
 
+void EnemyController::MoveTowardsPlayer() const
+{
+	if (current_target == nullptr) return;
+
+	float3 velocity = current_target->owner->transform.GetTranslation() - owner->transform.GetTranslation();
```
Comment (mariofv) [inline Scripts/src/Script/EnemyController.cpp:311]: Sme comment as before

File: Scripts/src/Script/EnemyController.cpp [hunk 7]
```diff
@@ -268,13 +321,16 @@ void EnemyController::LeaveAttackRange()
...
 	leave_attack_range = true;
 }
+#pragma endregion
 
-bool EnemyController::PlayerInSight()
+#pragma region Helper Methods
+bool EnemyController::PlayerInSight() const
 {
-	return (current_target == nullptr) ? false : current_target->owner->transform.GetTranslation().Distance(owner->transform.GetTranslation()) < detect_player_distance;
+	float3 target = current_target->owner->transform.GetTranslation();
```
Comment (mariofv) [inline Scripts/src/Script/EnemyController.cpp:329]: GlobalTranslation instead of local

File: Scripts/src/Script/Biter.cpp (hunk 3 above)
Comment (JoanStinson) [reply Scripts/src/Script/Biter.cpp:1]: ok

File: Scripts/src/Script/EnemyController.cpp (hunk 4 above)
Comment (JoanStinson) [reply Scripts/src/Script/EnemyController.cpp:235]: Yes, to follow the player

File: Scripts/src/Script/EnemyController.cpp (hunk 5 above)
Comment (JoanStinson) [reply Scripts/src/Script/EnemyController.cpp:287]: Because the y value is of the owner, not the current_target->owner

**Comment by mariofv** [comment]
Hey! I saw that ypu changed the GlobalTranslation thing, but be aware because there are some lines where you missed it. Example, you are here setting LocalTranslation instead Global:
```
owner->transform.SetTranslation(game_camera->transform.GetGlobalTranslation());
```

**Comment by JoanStinson** [comment]
> Hey! I saw that you changed the GlobalTranslation thing, but be aware because there are some lines where you missed it. Example, you are here setting LocalTranslation instead Global:
> 
> ```
> owner->transform.SetTranslation(game_camera->transform.GetGlobalTranslation());
> ```

ah lol I did cntr+f maybe I missed it, thank you! :))
